rust/src/state_machine/mod.rs

```diff
@@ -0,0 +1,311 @@
+use crate::{
+    coordinator::{MaskDict, RoundFailed, RoundSeed},
+    crypto::ByteObject,
+    mask::{Aggregation, BoundType, DataType, GroupType, MaskConfig, ModelType},
+    state_machine::{error::Error, idle::Idle, sum::Sum, sum2::Sum2, update::Update},
+    CoordinatorPublicKey,
+    CoordinatorSecretKey,
+    InitError,
+    PetError,
+    SeedDict,
+    SumDict,
+};
+use std::default::Default;
+use thiserror::Error;
+use tokio::sync::{mpsc, oneshot};
+
+mod error;
+mod idle;
+pub mod requests;
+mod sum;
+mod sum2;
+mod update;
+
+use requests::Request;
+
+#[derive(Debug)]
+pub struct CoordinatorState {
+    pk: CoordinatorPublicKey, // 32 bytes
+    sk: CoordinatorSecretKey, // 32 bytes
+
+    // round parameters
+    sum: f64,
+    update: f64,
+    seed: RoundSeed,
+    min_sum: usize,
+    min_update: usize,
+
+    // round dictionaries
+    /// Dictionary built during the sum phase.
+    sum_dict: SumDict,
```

Comment:
The dictionaries are kind of specific to each phase, so I'm wondering if we should move them to the phase's state:

```rust
pub struct Sum {
    sum_dict: SumDict,
}

pub struct Update {
    frozen_sum_dict: SumDict,
    seed_dict: SeedDict,
}

pub struct Sum2Phase {
    frozen_sum_dict: SumDict,
    frozen_seed_dict: SeedDict,
    mask_dict: MaskDict,
}
```

If we have an aggregation phase in the future, after the sum2 phase, we could also move the `aggregation` attribute out of `CoordinatorState`.

Reply:
moving the dicts into state specific data is a good idea in general, but we have to be careful here, for example the `sum_dict` is modified in the `Sum` and `Sum2` state, but the `Update` state of course comes in between, so we would need a way to transfer it.